### PR TITLE
Head and meta tags handling

### PR DESCRIPTION
**Pull request by revanth0212**
Added wrappers around the `HeadProvider` and `Title` components of `react-head`. Also exporting `Meta`, `Link` and `Style` tags directly from `react-head`.

These 5 components will be used to handle the document's `head` tag.

To be able to use the components anywhere in the app, the app has to be wrapped inside `HeadProvider` which provides the context needed to maintain all the rendered tags.

Closes [422](https://github.com/magento/pwa-studio/issues/422)

### CODE REVIEW

File: packages/venia-concept/src/components/Head/headProvider.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import React, { useEffect } from 'react';
+import { HeadProvider as _HeadProvider } from 'react-head';
+
+const filterOutTitleTags = titleTags => {
+    titleTags
+        .filter(titleNode => !titleNode.dataset.synteticTag)
```
Comment (brendanfalkowski) [inline packages/venia-concept/src/components/Head/headProvider.js]: synteticTag or syntheticTag?

File: packages/venia-concept/src/components/Head/title.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import { Title as _Title } from 'react-head';
+
+export default props => (
+    <_Title {...props} data-syntetictag>
```
Comment (brendanfalkowski) [inline packages/venia-concept/src/components/Head/title.js]: syntetictag or synthetictag?

File: packages/venia-concept/src/components/Head/headProvider.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        .forEach(titleTag => {
+            if (titleTag) titleTag.parentNode.removeChild(titleTag);
```
Comment (supernova-at) [inline packages/venia-concept/src/components/Head/headProvider.js]: Did this pass `prettier` formatting?

`yarn run prettier`

File: packages/venia-concept/src/components/Head/headProvider.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        });
+};
+
+export default props => {
```
Comment (supernova-at) [inline packages/venia-concept/src/components/Head/headProvider.js]: For consistency in the codebase let's give this function a semantic name and export it separately.

```js
const HeadProvider = props => { ... }

export default HeadProvider;
```

File: packages/venia-concept/src/components/Head/headProvider.js (hunk 1 above)
Comment (revanth0212) [reply packages/venia-concept/src/components/Head/headProvider.js]: My bad. Spelling mistake.

Comment (supernova-at) [reply packages/venia-concept/src/components/Head/headProvider.js]: Can you drop a comment here to explain why we're removing some and not all `title` tags, and how `syntheticTag` plays into that?

File: packages/venia-concept/src/components/Head/headProvider.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import React, { useEffect } from 'react';
+import { HeadProvider as _HeadProvider } from 'react-head';
```
Comment (supernova-at) [inline packages/venia-concept/src/components/Head/headProvider.js]: I'm not a huge fan of the underscore as the first character of a React component 😬

File: packages/venia-concept/package.json [hunk 6]
```diff
@@ -73,6 +74,7 @@
     "@magento/pwa-buildpack": "~3.0.0",
     "@magento/upward-js": "~2.2.0",
     "@storybook/react": "~4.1.13",
+    "@testing-library/react": "~8.0.8",
```
Comment (revanth0212) [inline packages/venia-concept/package.json]: Using `@testing-library/react` for DOM testing since the test renderer that comes with react renders components all in react as JS objects. Needed something like JSDom to assert head tags. Let me know if we can achieve this without `@testing-library/react`, will be happy to change.

FYI, the API `@testing-library/react` provides is very intuitive and can be used for other tests in the future.

File: packages/venia-concept/src/components/Head/title.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import { Title as _Title } from 'react-head';
+
+export default props => (
+    <_Title {...props} data-synthetictag>
```
Comment (supernova-at) [inline packages/venia-concept/src/components/Head/title.js]: Oh, so we're adding the synthetic tag? 

Why does `HeadProvider` need to filter them out then? How could they exist before `HeadProvider` is mounted?

File: packages/venia-concept/src/components/Head/title.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+        {props.children}
+    </_Title>
```
Comment (supernova-at) [inline packages/venia-concept/src/components/Head/title.js]: Still not a fan of the underscore 😝

File: packages/venia-concept/src/components/Head/title.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import { Title as _Title } from 'react-head';
+
+export default props => (
```
Comment (supernova-at) [inline packages/venia-concept/src/components/Head/title.js]: For consistency in the codebase please also define and export this function separately.

File: packages/venia-concept/package.json (hunk 6 above)
Comment (supernova-at) [reply packages/venia-concept/package.json]: Oh, I assumed @jimbo steered you in this direction. We probably want his sign off on this one.

**Comment by supernova-at** [comment]
https://github.com/magento/pwa-studio/pull/1537#issuecomment-519243093. Please fill out the PR template :)

Looks like there are some conflicts to resolve as well 👍

**Comment by supernova-at** [comment]
I also have a small concern about stale meta tags hanging around.

If Page2 provides the same meta tag as Page1 with a different value, it'll get overwritten properly. But what if Page1 has a specific meta tag that Page2 doesn't care / know about? Won't that just stick around forever?

File: packages/venia-concept/src/components/Head/headProvider.js (hunk 1 above)
Comment (revanth0212) [reply packages/venia-concept/src/components/Head/headProvider.js]: So the problem with `react-head` is, it won't render a `title` tag if there is one already present in the markup. So according to their docs, it is required to remove the existing tags and then render the needed ones using `react-head`. Now we should not be removing the tags that were created using `react-head` in case of a re-render. That is where the `synthetictag` flag comes in. If the `syntheticTag` flag is set, it means that title tag was created using `react-head` and should not be removed.

File: packages/venia-concept/src/components/Head/headProvider.js (hunk 3 above)
Comment (revanth0212) [reply packages/venia-concept/src/components/Head/headProvider.js]: Yeah, it did. Does it mean I am running a different prettier config than yours?

File: packages/venia-concept/src/components/Head/title.js (hunk 7 above)
Comment (revanth0212) [reply packages/venia-concept/src/components/Head/title.js]: Explained in the above comment.

File: packages/venia-concept/src/components/Head/headProvider.js (hunk 5 above)
Comment (revanth0212) [reply packages/venia-concept/src/components/Head/headProvider.js]: What shall I change it to? I wanted to signify vanila `HeadProvider` as` _HeadProvider`.

**Comment by revanth0212** [comment]
> I also have a small concern about stale meta tags hanging around.
> 
> If Page2 provides the same meta tag as Page1 with a different value, it'll get overwritten properly. But what if Page1 has a specific meta tag that Page2 doesn't care / know about? Won't that just stick around forever?

If the new page is a new HTML it shouldn't unless the new markup has that meta tag. If it a page transition using react-router, then it is on us to clear those tags. As of today, `react-head` does not support the manipulations of tags (like removing). I have checked their codebase, it is possible for us to go ahead and add manipulations API as submit a PR, or the simplest would be to write a simple wrapper around `Meta` (just like how we did for `Title`) that does these kinds of manipulations.

The ticket dint mention those use cases but you are right, they will be hanging around and there has to be a way to delete the existing ones. Shall I go ahead and write that wrapper? What do you say @jimbo @supernova-at?

File: packages/venia-concept/src/components/Head/__tests__/headProvider.spec.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+import { render, cleanup } from '@testing-library/react';
+import TestRenderer from 'react-test-renderer';
+
+import { HeadProvider, Title, Link, Meta, Style } from '../';
+
+afterEach(cleanup);
+
+test('snapshot testing HeadProvider', () => {
```
[the remark this answers is not in the document]
Comment (revanth0212) [reply packages/venia-concept/src/components/Head/__tests__/headProvider.spec.js]: Not sure what to call this. It is a casual snapshot test.

File: packages/venia-concept/package.json (hunk 6 above)
Comment (revanth0212) [reply packages/venia-concept/package.json]: This is on me. Jimmy dint ask me to use this package. But if you guys have any other solution to render components on JSDOM or similar DOM implementations without a new library, let me know.

File: packages/venia-concept/src/components/Head/__tests__/headProvider.spec.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (supernova-at) [reply packages/venia-concept/src/components/Head/__tests__/headProvider.spec.js]: "HeadProvider by itself doesn't render anything" ?

Basically you just want someone who is only looking at the snapshot to be able to ensure it is correct.

File: packages/venia-concept/src/components/Head/headProvider.js (hunk 3 above)
Comment (supernova-at) [reply packages/venia-concept/src/components/Head/headProvider.js]: No, you're right - it worked for me too. Just haven't seen very many of these I guess 🤷‍♂

**Comment by supernova-at** [comment]
> If it a page transition using react-router, then it is on us to clear those tags.

That is the case when navigating within the app, yes.

> Shall I go ahead and write that wrapper?

I think we do need it.

File: packages/venia-concept/src/components/Head/headProvider.js (hunk 1 above)
Comment (supernova-at) [reply packages/venia-concept/src/components/Head/headProvider.js]: > the problem with react-head is, it won't render a title tag if there is one already present in the markup.

Isn't that what [this test](https://github.com/magento/pwa-studio/pull/1537/files#diff-6ad3f434d87e531558aa3b745f33c3ecR40) does though? Or does that test only work because you added the synthetic tag handling?

Comment (supernova-at) [reply packages/venia-concept/src/components/Head/headProvider.js]: I meant drop a comment in the code itself so that in the future no one removes this line thinking that it's not needed 😛

File: packages/venia-concept/package.json (hunk 6 above)
Comment (supernova-at) [reply packages/venia-concept/package.json]: I think this is the recommended approach, I was just surprised we didn't already have it. I'm 👍

File: packages/venia-concept/src/components/Head/headProvider.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        .forEach(titleTag => {
+            if (titleTag) titleTag.parentNode.removeChild(titleTag);
+        });
+};
+
+const HeadProvider = props => {
+    useEffect(() => {
+        const titleTagsToRemove = document.getElementsByTagName('title');
+        filterOutTitleTags([...titleTagsToRemove]);
+    }, []);
```
Comment (supernova-at) [inline packages/venia-concept/src/components/Head/headProvider.js]: I'm still hoping we can simplify this 😄 

Because we're passing an empty array as the second argument to `useEffect`, it will run on mount and unmount only:

> If you want to run an effect and clean it up only once (on mount and unmount), you can pass an empty array ([]) as a second argument. This tells React that your effect doesn’t depend on any values from props or state, so it never needs to re-run. This isn’t handled as a special case — it follows directly from how the dependencies array always works.

Since our `Title` component doesn't do anything unless it's wrapped in this `HeadProvider` component, there can never be a case where the `Title` component has inserted a `<title>` that the `HeadProvider` wants to skip removing, right?

If I'm understanding all that correctly, then there is no need for the synthetic tag data attribute, and therefore no need to filter for them here - just removing all `<title>` tags should be sufficient 👍

Comment (revanth0212) [reply packages/venia-concept/src/components/Head/headProvider.js]: You are right. But that was added in place to handle the server-rendered output. If the server-rendered output already has a `synthetictag` on it, we don't want to delete it.

But after talking to you, I have realized that is a bleak use-case. We can probably get rid of the filtering for now and add this logic when needed.

File: packages/venia-concept/src/components/Head/headProvider.js (hunk 5 above)
Comment (revanth0212) [reply packages/venia-concept/src/components/Head/headProvider.js]: Interesting. Let me get this done.

Comment (brendanfalkowski) [reply packages/venia-concept/src/components/Head/headProvider.js]: @supernova-at The pattern of naming things `VeniaHeadProvider` after the package is kind of frustrating in practice. These components get used in implementations that aren't identified as Venia anymore and that naming doesn't feel good to see, so we rewrite and maintain that rewrite. It makes upstream integration more complex.

**Comment by revanth0212** [comment]
> Looks straightforward. Remove `@testing-library/react` and fix `yarn.lock` and this is good to go. 👍

Done. Refactored tests to not use that library.

**Comment by dpatil-magento** [comment]
@revanth0212  I don't see original bug has fixed. 
1. Click on any category say - "Tops"
expected - Title should say "Tops - Venia"
actual - Title does not change , its still - "Home Page - Venia"

2. Go to ant PDP say /valeria-two-layer-tank.html
expected - title should say "Valeria Two-Layer Tank - Venia"
actual - "Home Page - Venia"
Note - After page reload - "Valeria Two-Layer Tank - "  `venia` text still missing.

**Comment by revanth0212** [comment]
Dev is right. We are able to reproduce this issue using his steps.

image

Will look into the issue.

File: packages/venia-ui/upward.yml [hunk 12]
```diff
@@ -100,6 +100,10 @@ veniaAppShell:
         urlResolver: veniaUrlResolver
         env: env
         mediaBackendURL: mediaBackendURL
+        site:
+          inline:
+            name:
+              inline: Venia
```
Comment (supernova-at) [inline packages/venia-ui/upward.yml]: Why are these changes needed? I'm assuming this is just changing the title to "Venia" whenever it is supplied by Upward?

Comment (supernova-at) [reply packages/venia-ui/upward.yml]: Did you not also have to change any part of the templates themselves?

Comment (revanth0212) [reply packages/venia-ui/upward.yml]: This is needed for the product template that needs `site.name`. Others don't as of now. I can create a new PR that changes other templates as well to use `site.name`.

**Comment by supernova-at** [comment]
This looks great! One thing I noticed while navigating:

Sometimes while it is loading a new route, the title of the tab changes to `Home Page - Venia`. If we can't get that out of there, might we consider changing the title in `app.js` to just be "Venia"?

**Comment by revanth0212** [comment]
> This looks great! One thing I noticed while navigating:
> 
> Sometimes while it is loading a new route, the title of the tab changes to `Home Page - Venia`. If we can't get that out of there, might we consider changing the title in `app.js` to just be "Venia"?

Sounds good to me, but there was a reason why I went with `Home Page - Venia`. The UPWARD server outputs HTML that has `Home page - Venia` (small P), so if we don't have the same title, for a brief second you'll see `Home page - Venia` from UPWARD and then once our app loads it will change to `Venia` and then to the product's title. 

We will also need to change the UPWARD server's output for the home page to have `P` not `p` to be consistent.

**Comment by dpatil-magento** [comment]
@revanth0212 Please check unit test failures. Able to reproduce on my local as well.
    packages/venia-ui/lib/components/App/__tests__/app.spec.js
    packages/venia-ui/lib/RootComponents/Category/__tests__/categoryContent.spec.js

**Comment by jimbo** [comment]
This is ready to go pending QA approval.
